src/x86_64/kernlock.c

```diff
@@ -0,0 +1,63 @@
+#include <runtime.h>
+#include <x86_64.h>
+#include <apic.h>
+
+static struct {
+    u64 size;
+    u64 sem;
+    u64 head;                   /* next to enqueue */
+    u64 tail;                   /* next to dequeue */
+} kernlock;
+
+static u64 *kernlock_ring;
+
+static inline int dequeue_cpu(void)
+{
+    u64 src = fetch_and_add(&kernlock.tail, 1);
+    return (int)kernlock_ring[src % kernlock.size];
+}
+
+static inline void enqueue_cpu(int cpu)
+{
+    u64 dest = fetch_and_add(&kernlock.head, 1);
+    kernlock_ring[dest % kernlock.size] = cpu; /* cas for paranoia? */
```

Comment:
this has a race...it matches the original queue implementation that ... Levent I guess sort of made worse. 

otherwise the actual locked queue is fine. it seems like its a specialized version of something we will need more of? so I guess we could allocate it as an object rather than a global?

idk..maybe just a more primitive queue that's thread safe that we can use to implement queues of whatever?

Reply:
yeah. thinking about it. this is really our mutex implementation, except that as we drop it into the more general part of the kernel it will be queuing closures instead of cpus...and if we fixed up the other queue (runqueue and I guess bhqueue?) that could probably be merged as well. not as sure about all the stuff in the block queue?

Reply:
Not sure where the race is on the enqueue. There's no check for a semaphore maximum because it is assumed that the ring is sized to the maximum number of cpus, and thus cannot overflow.

Granted, it's possible that one cpu could bump the semaphore and that another cpu could immediately after bump the semaphore and also grab the next slot on the ring first. But as the advancing of the head is atomic, the first cpu would just get enqueued second. At first I thought this would be a problem, but it's just very close contention between two cpus - does it matter? It's not a priority inversion, right?

I think someone tried to make the generic queue safe for multiple readers as well...but that side is clearly broken. Here, we don't have multiple readers as there is only one lock holder at a time. I figured it would be cheaper to just implement it within rather than going to the general queue with the heavy dequeue, even if fixed.

Or maybe I'm totally missing something else?

The blockq is just a unix-specific thing to contain thread waiters on a resource. It could use a ring queue, but I just didn't want to deal with the hard upper limit. It may at some point need finer-grained locking around it if we get to that level of granularity in the kernel locking in general.

Reply:
yeah, advancing the head is atomic, so there isn't any race between competing writers, they each get a unique and consecutive storage slot. the only issue is a race between the reader and a writer, since the write to the slot isn't atomic wrt the pointer advance.

so the reader could see that there is something on the queue, read it, and get back the old contents before the writer has a chance to update the queue entry

Reply:
Thanks...don't know how I missed that. I'll take this back to the drawing board - and maybe remove it for this PR.

Reply:
it is a pretty* small race. this is the one the levent was tasked with closing. if we know that the queue entry are always pointers that we have control over (not random ints, or user specified pointers), then there are plenty of bits to play with to mark invalid entries (which would require a spin to detect? I guess there is still an issue with progress if the writer took an interrupt right there (the only time this would really happen))...right, just INVALID pointer after the read and before advancing the read pointer would be enough here if we could tolerate the spin